libkmod/libkmod-builtin.c

```diff
@@ -1,333 +1,193 @@
+			if (count == 0)
+				continue;
 			break;
 		}
-
-		iter->nstrings++;
-		pos = offset;
+		if (!strbuf_pushchars(buf, dot + 1) ||
+		    !strbuf_pushchar(buf, '\0')) {
+			count = -errno;
+			ERR(info->ctx, "get_strings: "
+			    "failed to append modinfo string\n");
+			return count;
+		}
+		count++;
 	}
 
-	iter->pos = iter->next;
-	iter->next = pos;
-
-	free(modname);
+	if (count == INTPTR_MAX) {
+		count = -ENOMEM;
+		ERR(info->ctx, "get_strings: "
+		    "too many modinfo strings encountered\n");
+		return count;
+	}
 
-	return (iter->pos < iter->size);
+	return count;
 }
 
-static bool kmod_builtin_iter_get_modname(struct kmod_builtin_iter *iter,
-				char modname[static PATH_MAX])
+static char **strbuf_to_vector(struct strbuf *buf, size_t count)
 {
-	int sv_errno;
-	char *line, *dot;
-	size_t linesz, len;
-	off_t offset;
-
-	if (iter->pos == iter->size) {
-		sv_errno = EINVAL;
-		goto fail;
-	}
-
-	line = NULL;
+	/* size required for string vector + terminating NULL */
+	const size_t vecsz = sizeof(char *) * (count + 1);
+	char **vector;
+	char *s;
+	size_t n;
 
-	offset = get_string(iter, iter->pos, &line, &linesz);
-	if (offset <= 0) {
-		sv_errno = errno;
-		if (offset)
-			ERR(iter->ctx, "get_string: %s\n", strerror(errno));
-		goto fail;
+	/* make sure that vector and strings fit into memory constraints */
+	if (SIZE_MAX / sizeof(char *) - 1 < count ||
+	    SIZE_MAX - buf->used < vecsz) {
+		errno = ENOMEM;
+		return NULL;
 	}
 
-	dot = strchr(line, '.');
-	if (!dot) {
-		sv_errno = EINVAL;
-		ERR(iter->ctx, "kmod_builtin_iter_get_modname: unexpected string without modname prefix\n");
-		goto fail;
-	}
-
-	len = dot - line;
+	vector = realloc(buf->bytes, vecsz + buf->used);
+	if (vector == NULL)
+		return NULL;
+	buf->bytes = NULL;
+	memmove(vector + count + 1, vector, buf->used);
+	s = (char *)(vector + count + 1);
 
-	if (len >= PATH_MAX) {
-		sv_errno = ENAMETOOLONG;
-		goto fail;
+	for (n = 0; n < count; n++) {
+		vector[n] = s;
+		s += strlen(s) + 1;
 	}
+	vector[n] = NULL;
 
-	strncpy(modname, line, len);
-	modname[len] = '\0';
-
-	return true;
-fail:
-	errno = sv_errno;
-	return false;
+	return vector;
 }
 
 /* array will be allocated with strings in a single malloc, just free *array */
 ssize_t kmod_builtin_get_modinfo(struct kmod_ctx *ctx, const char *modname,
 				char ***modinfo)
 {
-	ssize_t count = 0;
-	char *s, *line = NULL;
-	size_t i, n, linesz, modlen, size;
-	off_t pos, offset;
+	struct kmod_builtin_info info;
+	struct strbuf buf;
+	ssize_t count;
 
-	char *name = NULL;
-	char buf[PATH_MAX];
-
-	struct kmod_builtin_iter *iter = kmod_builtin_iter_new(ctx);
-
-	if (!iter)
+	if (!kmod_builtin_info_init(&info, ctx))
 		return -errno;
-
-	while (!name && kmod_builtin_iter_next(iter)) {
-		if (!kmod_builtin_iter_get_modname(iter, buf)) {
+	strbuf_init(&buf);
+
+	count = get_strings(&info, modname, &buf);
+	if (count == 0)
+		*modinfo = NULL;
+	else if (count > 0) {
+		*modinfo = strbuf_to_vector(&buf, (size_t)count);
```

Comment:
I'd consider strbuf_to_vector() as a function stealing/disowning the buffer. As such it's odd to still call strbuf_release() below.

Reply:
If everything works as expected, then yes: `strbuf_to_vector` would steal/disown the buffer. But if `realloc` fails, then the `strbuf` is considered "left alone for caller to clean up who passed reference to us."

Going the Rust way, we could argue that we pass ownership to `strbuf_to_vector`. But on the other hand, from a C perspective, we construct the `strbuf` in `kmod_builtin_get_modinfo` so it should be its responsibility to clean up any resources associated with it when leaving.

So I went with the second (C) approach. Let me know what you think, I don't have a hard opinion either way.

Reply:
I've moved the cleanup into error handling. I think it actually looks nicer this way, like "I expect this function to convert the buffer and only if it fails, I release the whole thing on my own."